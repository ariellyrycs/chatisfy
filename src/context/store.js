import React, { useReducer, createContext, useContext } from 'react';
import Mock from '../Mock';
import { generateUId, generateTime } from '../Utils/General.js';


const currentUser = {
	id: generateUId(),
	nickName: ''
};

const Context = createContext();

const useAppContext = () => useContext(Context);

const initialState = {
	...Mock,
	currentUser
};

const mainReduce = (state, action) => {
	switch (action.type) {
		case 'UPDATE_ROOM':
			return {
				...state,
				selectedRoom: action.roomId
			};
		case 'SEND_MESSAGE':
			let currentId = generateUId();
			let rooms = state.rooms;
			let messageByRooms = rooms.messages;
			let messages = state.messages;
			let currentRoomId = state.selectedRoom,
				sentBy = action.sentBy;
			return {
				...state,
				rooms: {
					...rooms,
					messages: {
						...rooms.messages,
						[currentRoomId]: [...(messageByRooms[currentRoomId] || []), currentId]
					}
				},
				messages: {
					...messages,
					data: {
						...messages.data,
						[currentId]: {
							id: currentId,
							from: sentBy,
							content: action.message,
							createdTime: generateTime(),
							seenBy: [],
							own: action.from === sentBy
						}
					}
				}
			};
		case 'UPDATE_NICKNAME':
			return {
				...state,
				currentUser: {
					...state.currentUser,
					nickName: action.nickName
				}
			};
		default:
			return state;
	}
};

function ContextProvider(props) {
	const [state, dispatch] = useReducer(mainReduce, initialState);
	return (
		<Context.Provider value={{ state, dispatch }}>
			{props.children}
		</Context.Provider>
	);
};

export { ContextProvider, Context, useAppContext };