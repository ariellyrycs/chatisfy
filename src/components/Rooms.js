import React from 'react';

import {useAppContext} from '../context/store';

import {
    ListItem,
    ListItemText,
    List,
    Divider} from '@material-ui/core';


export default function Rooms({handleDrawerClose}) {

    const {state, dispatch} = useAppContext();

    const rooms = Object.values(state.rooms.data);

    const selectedRoom = state.selectedRoom;

    const transitionToRoom = (id) => {
        dispatch({type: 'UPDATE_ROOM', roomId: id})
        handleDrawerClose();
    };
    
    return (
        <>
            <ListItem>
                <ListItemText primary={state.currentUser.nickName} />
            </ListItem>
            <Divider />
            <List component="nav" aria-label="Rooms">
                {rooms.map((room) => (
                    <ListItem
                        key={room.id}
                        className='list-item'
                        selected={selectedRoom === room.id}
                        button
                        onClick={() => transitionToRoom(room.id)}>
                        <ListItemText primary={room.name} />
                    </ListItem>
                ))}
            </List>
        </>
    )
}