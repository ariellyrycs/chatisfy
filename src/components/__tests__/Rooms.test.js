import React from 'react';
import Rooms from '../Rooms';
import * as Store from '../../context/store';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'

describe('<Rooms />', () => {

    test('trigget distpatch with the right params', () => {
        let handleDrawerCloseStub = jest.fn();
        let dispatchStub = jest.fn();
        let contextValues = {
            state: {
                rooms: {
                    data: {
                        1: {
                            id: 1,
                            from: 1,
                            content: 'Make sure you don\'t, as they say, go whole',
                            createdTime: '14:23',
                            seen: false
                        },
                    }
                },
                currentUser: {
                    nickName: 'Pedro'
                }
            },
            dispatch: dispatchStub
        };
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<Rooms handleDrawerClose={handleDrawerCloseStub}/>);
        tree.find('.list-item').first().simulate('click');
        expect(dispatchStub).toHaveBeenCalledWith(
            expect.objectContaining({type: 'UPDATE_ROOM', roomId: 1})
        );
        expect(handleDrawerCloseStub).toHaveBeenCalled();
    });

    test('display component correctly', function () {
        let contextValues = {
            state: {
                rooms: {
                    data: {
                        1: {
                            id: 1,
                            from: 1,
                            content: 'Make sure you don\'t, as they say, go whole',
                            createdTime: '14:23',
                            seen: false
                        },
                    }
                },
                currentUser: {
                    nickName: 'Pedro'
                }
            },
            dispatch: () => {}
        };
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<Rooms handleDrawerClose={() => {}}/>);
        expect(toJson(tree)).toMatchSnapshot();
    });
});
