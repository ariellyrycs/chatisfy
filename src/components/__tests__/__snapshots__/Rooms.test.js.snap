// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Rooms /> display component correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))>
    <WithStyles(ForwardRef(ListItemText))
      primary="Pedro"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Divider)) />
  <WithStyles(ForwardRef(List))
    aria-label="Rooms"
    component="nav"
  >
    <WithStyles(ForwardRef(ListItem))
      button={true}
      className="list-item"
      key="1"
      onClick={[Function]}
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemText)) />
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
</Fragment>
`;
